code/dropout.py

```diff
@@ -0,0 +1,447 @@
+        :param input: a symbolic tensor of shape (n_examples, n_in)
+
+        :type n_in: int
+        :param n_in: dimensionality of input
+
+        :type n_out: int
+        :param n_out: number of hidden units
+
+        :type activation: theano.Op or function
+        :param activation: Non linearity to be applied in the hidden
+                           layer
+        :type dropout_rate: list 
+        :param dropout_rate: array containing probabilities of retaining a unit
+        """
+        super(HiddenDropoutLayer, self).__init__(
+                rng=rng, input=input, n_in=n_in, n_out=n_out, W=W, b=b,
+                activation=activation)
+
+        self.output = _dropsout(rng, self.output, p=dropout_rate)
+    
+
+class MLP(object):
+    """Multi-Layer Perceptron Class
+
+    A multilayer perceptron is a feedforward artificial neural network model
+    that has one layer or more of hidden units and nonlinear activations.
+    Intermediate layers usually have as activation function tanh or the
+    sigmoid function (defined here by a ``HiddenLayer`` class)  while the
+    top layer is a softmax layer (defined here by a ``LogisticRegression``
+    class).
+    """
+
+    def __init__(self, rng, input, n_in, n_hidden, dropout_rates, n_out):
+        """Initialize the parameters for the multilayer perceptron
+
+        :type rng: numpy.random.RandomState
+        :param rng: a random number generator used to initialize weights
+
+        :type input: theano.tensor.TensorType
+        :param input: symbolic variable that describes the input of the
+        architecture (one minibatch)
+
+        :type n_in: int
+        :param n_in: number of input units, the dimension of the space in
+        which the datapoints lie
+
+        :type n_hidden: int
+        :param n_hidden: number of hidden units
+
+        :type dropout_rate: list 
+        :param dropout_rate: array containing probabilities of retaining a unit
+
+        :type n_out: int
+        :param n_out: number of output units, the dimension of the space in
+        which the labels lie
+
+        """
+        
+        #Dropping out the input layer
+        inp_dropout_layer = _dropsout(rng, input, p=dropout_rates[0])
+        
+        self.drop_layer = HiddenDropoutLayer(rng=rng,
+                    input=inp_dropout_layer,
+                    activation=T.tanh,
+                    n_in=n_in, n_out=n_hidden,
+                    dropout_rate=dropout_rates[1])
+        
+
+        # Since we are dealing with a one hidden layer MLP, this will translate
+        # into a HiddenLayer with a tanh activation function connected to the
+        # LogisticRegression layer; the activation function can be replaced by
+        # sigmoid or any other nonlinear function
+        self.hiddenLayer = HiddenLayer(
+            rng=rng,
+            input=input,
+            n_in=n_in,
+            n_out=n_hidden,
+            W=self.drop_layer.W * (1 - dropout_rates[1]),
```

Comment:
Oh, so you scale the weights down when not using dropout rather than of upscaling the inputs after the dropout.
I guess it is more or less equivalent, and I don't really have much experience with dropout, but I would guess that keeping the magnitude of the input independent of the dropout factor would be better.
Also, it would be more efficient to scale the input rather than the weights.